### ISSUE TITLE
Initialization with zero - "missing metrics problem"

### ISSUE THREAD

**Issue by stempler** [martin-helmich/prometheus-nginxlog-exporter#130]
**Is your feature request related to a problem? Please describe.**
Currently metrics like the response counter and the histograms are not initialized with zero values.
Doing the initialization to zero is challenging, as you would have to know all possible label combinations beforehand (at least one scrape before there is actual data), which may be easy for some labels (e.g. methods) but hard for others where it is not clear which values they might take.

The problematic with metrics that are not initialized to zero is described for instance [here](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics) and [here](https://www.robustperception.io/existential-issues-with-metrics), but let's give a practical example:

Let's say we have a scrape interval of `15s`. We have a graph or alert that uses `increase(...)` or `rate(...)` to inform us about requests with status `500` based on the response counter metric.

Between two scrapes for the first time there are requests w/ status `500`. That means when the time series first appears in a scrape, the related request counter is already set to the number of requests w/ status `500` that occurred in the meantime, let's say it were `50` requests between those two scrapes.
Missing a previous zero value, `increase` or `rate` cannot detect a change and thus these 50 requests go unnoticed. Similar requests that happen after the metric was initialized will show up, but if there are none, there won't be any trace at all of `500`s on the graph.

As it is now, at least the first request in a counter "gets lost" like this, as it will never initialize below `1`, though the number is likely usually higher.

**Describe the solution you'd like**
I would be mainly interested in your thoughts on this, and if you maybe see a workaround.
For a solution, I currently don't see any alternatives to attempting an initialization based on expected label combinations.

**Comment by stempler** [comment]
A possible way to deal with this (though it is not pretty) that comes to my mind is the following:

1. When a metric (with a specific combination of label values) is scraped for the first time, instead of the current value yield zero together with a timestamp that lies in the past (e.g. the time of the previous scrape where it was not collected yet)
2. On any subsequent scrapes/collections just yield the current value as usual

That way the initialization to zero would happen, even if somewhat late, and we would loose the information of the actual count/sum from the initial scrape, but increase/rate calculations would be correct.

**Comment by martin-helmich** [comment]
This is a tough one. 😓  I recognize the issue, but as you've already stated, it's quite impossible to determine all possible combinations of label values beforehand.

Retroactively providing zero-metrics upon the first scrape also seems suboptimal to me, since the exporter would need to be able to actually determine when a metric was scraped for the first time. Also, we'd need to export values for the same metric with different timestamps (which is probably not the most common use case of the instrumentation library).

I'll need some time to give this some more tought. 🤔  Stay tuned -- and of course, any further input on this matter is welcome.

**Comment by stempler** [comment]
I implemented a proof of concept for the approach I sketched above and created an issue asking if this is something that might be considered to be added to the client library (as I assume that would be a requirement for you to consider using that functionality): https://github.com/prometheus/client_golang/issues/780

**Comment by stempler** [comment]
In https://github.com/prometheus/client_golang/issues/780 there is an explanation why the approach I sketched can't be a general solution to that problem (e.g. when using multiple Prometheus instances to scrape) and thus is not considered for a PR. For my use case it seems to work well enough. Even if there is no guarantee that the metrics will be initialized with zero, the situation is definitely an improvement.

If anyone is interested I have the changes on these forks/branches:

Prometheus client:
https://github.com/stempler/client_golang/tree/exp/counter-initialization

Exporter:
https://github.com/wetransform-os/prometheus-nginxlog-exporter/tree/belate-init (mainly https://github.com/wetransform-os/prometheus-nginxlog-exporter/commit/fd4d289982664c7352081ba62bba9fed6a3f050a)

Right now there is at least the problem that the `ConstLabels` in the metrics are not properly supported with my changes (and thus using a label for the namespace. Otherwise I did not observe any problems so far.
